File: plugins/memory/postgres-lcm/src/expand.ts
/**
 * LCM Expander — DAG traversal over the summary tree.
 *
 * Replicates lcm_expand and lcm_describe functionality.
 * Walks summary_parents downward, then summary_messages for leaf content.
 */

import pg from 'pg';

export interface SummaryNode {
  summaryId: string;
  conversationId: number;
  kind: string;
  depth: number;
  content: string;
  tokenCount: number;
  earliestAt: Date | null;
  latestAt: Date | null;
  descendantCount: number;
  createdAt: Date;
  model: string;
}

export interface ExpandResult {
  summary: SummaryNode;
  children: SummaryNode[];
  sourceMessages: Array<{ messageId: number; role: string; content: string; createdAt: Date }>;
}

export class LcmExpander {
  private pool: pg.Pool;

  constructor(pool: pg.Pool) {
    this.pool = pool;
  }

  /**
   * Describe a summary — metadata + lineage.
   */
  async describe(summaryId: string): Promise<SummaryNode | null> {
    const result = await this.pool.query(
      `SELECT summary_id, conversation_id, kind, depth, content, token_count,
              earliest_at, latest_at, descendant_count, created_at, model
       FROM summaries WHERE summary_id = $1`,
      [summaryId],
    );

    if (result.rows.length === 0) return null;
    return this.mapNode(result.rows[0]);
  }

  /**
   * Expand a summary — get children and source messages.
   */
  async expand(summaryId: string, maxDepth: number = 3): Promise<ExpandResult | null> {
    const summary = await this.describe(summaryId);
    if (!summary) return null;

    // Get child summaries
    const children = await this.getChildren(summaryId);

    // Get source messages (leaf nodes)
    const sourceMessages = await this.getSourceMessages(summaryId);

    return { summary, children, sourceMessages };
  }

  /**
   * Recursively expand to a given depth.
   */
  async expandDeep(summaryId: string, maxDepth: number = 3): Promise<ExpandResult | null> {
    const result = await this.expand(summaryId);
    if (!result || maxDepth <= 1) return result;

    // Recursively expand children
    for (const child of result.children) {
      const childExpanded = await this.expandDeep(child.summaryId, maxDepth - 1);
      if (childExpanded) {
        result.sourceMessages.push(...childExpanded.sourceMessages);
      }
    }

    return result;
  }

  /**
   * Get direct children of a summary.
   */
  private async getChildren(summaryId: string): Promise<SummaryNode[]> {
    const result = await this.pool.query(
      `SELECT s.summary_id, s.conversation_id, s.kind, s.depth, s.content, s.token_count,
              s.earliest_at, s.latest_at, s.descendant_count, s.created_at, s.model
       FROM summaries s
       JOIN summary_parents sp ON sp.summary_id = s.summary_id
       WHERE sp.parent_summary_id = $1
       ORDER BY sp.ordinal`,
      [summaryId],
    );

    return result.rows.map(this.mapNode);
  }

  /**
   * Get source messages linked to a summary.
   */
  private async getSourceMessages(
    summaryId: string,
  ): Promise<Array<{ messageId: number; role: string; content: string; createdAt: Date }>> {
    const result = await this.pool.query(
      `SELECT m.message_id, m.role, m.content, m.created_at
       FROM messages m
       JOIN summary_messages sm ON sm.message_id = m.message_id
       WHERE sm.summary_id = $1
       ORDER BY sm.ordinal`,
      [summaryId],
    );

    return result.rows.map((r) => ({
      messageId: r.message_id,
      role: r.role,
      content: r.content,
      createdAt: r.created_at,
    }));
  }

  private mapNode(row: any): SummaryNode {
    return {
      summaryId: row.summary_id,
      conversationId: row.conversation_id,
      kind: row.kind,
      depth: row.depth,
      content: row.content,
      tokenCount: row.token_count,
      earliestAt: row.earliest_at,
      latestAt: row.latest_at,
      descendantCount: row.descendant_count,
      createdAt: row.created_at,
      model: row.model,
    };
  }
}

File: integrations/hermes/rivet-memory/expand.py
"""DAG traversal over ``ros_summaries``.

Direct port of ``plugins/memory/postgres/src/expand.ts``. ``parent_id`` lives
on the summary row; source messages link through ``ros_summary_sources``.
"""

from __future__ import annotations

# See ``tools.py`` for the rationale behind this namespace bootstrap.
import sys as _sys
import types as _types

_top = __name__.split(".", 1)[0]
if _top.startswith("_") and _top not in _sys.modules:
    _sys.modules[_top] = _types.ModuleType(_top)

from dataclasses import dataclass
from datetime import datetime
from typing import List, Optional

from .client import RivetMemoryClient


@dataclass
class SummaryNode:
    summary_id: str
    conversation_id: Optional[str]
    kind: str
    depth: int
    content: str
    message_count: int
    earliest_at: Optional[datetime]
    latest_at: Optional[datetime]
    created_at: datetime
    model: Optional[str]
    access_count: int


@dataclass
class SourceMessage:
    message_id: str
    role: str
    content: str
    created_at: datetime


@dataclass
class ExpandResult:
    summary: SummaryNode
    children: List[SummaryNode]
    source_messages: List[SourceMessage]


_SUMMARY_COLUMNS = """
id, conversation_id, kind, depth, content, message_count,
earliest_at, latest_at, created_at, model, access_count
"""


def _row_to_summary(row: tuple) -> SummaryNode:
    return SummaryNode(
        summary_id=str(row[0]),
        conversation_id=str(row[1]) if row[1] is not None else None,
        kind=row[2],
        depth=row[3],
        content=row[4],
        message_count=row[5],
        earliest_at=row[6],
        latest_at=row[7],
        created_at=row[8],
        model=row[9],
        access_count=row[10] or 0,
    )


class Expander:
    def __init__(self, client: RivetMemoryClient) -> None:
        self._client = client

    def describe(self, summary_id: str) -> Optional[SummaryNode]:
        with self._client.connection() as conn:
            with conn.cursor() as cur:
                cur.execute(
                    f"SELECT {_SUMMARY_COLUMNS} FROM ros_summaries WHERE id = %s",
                    (summary_id,),
                )
                row = cur.fetchone()
        return _row_to_summary(row) if row else None

    def expand(self, summary_id: str) -> Optional[ExpandResult]:
        summary = self.describe(summary_id)
        if summary is None:
            return None
        children = self._get_children(summary_id)
        sources = self._get_source_messages(summary_id)
        return ExpandResult(summary=summary, children=children, source_messages=sources)

    def expand_deep(self, summary_id: str, max_depth: int = 3) -> Optional[ExpandResult]:
        result = self.expand(summary_id)
        if result is None or max_depth <= 1:
            return result
        for child in result.children:
            child_result = self.expand_deep(child.summary_id, max_depth - 1)
            if child_result is not None:
                result.source_messages.extend(child_result.source_messages)
        return result

    # -- Internal -----------------------------------------------------------

    def _get_children(self, summary_id: str) -> List[SummaryNode]:
        with self._client.connection() as conn:
            with conn.cursor() as cur:
                cur.execute(
                    f"SELECT {_SUMMARY_COLUMNS} FROM ros_summaries "
                    "WHERE parent_id = %s ORDER BY created_at",
                    (summary_id,),
                )
                rows = cur.fetchall()
        return [_row_to_summary(r) for r in rows]

    def _get_source_messages(self, summary_id: str) -> List[SourceMessage]:
        with self._client.connection() as conn:
            with conn.cursor() as cur:
                cur.execute(
                    """
                    SELECT m.id, m.role, m.content, m.created_at
                      FROM ros_messages m
                      JOIN ros_summary_sources ss ON ss.message_id = m.id
                     WHERE ss.summary_id = %s
                     ORDER BY ss.ordinal
                    """,
                    (summary_id,),
                )
                rows = cur.fetchall()
        return [
            SourceMessage(
                message_id=str(r[0]),
                role=r[1],
                content=r[2] or "",
                created_at=r[3],
            )
            for r in rows
        ]
